### ISSUE TITLE
Some data operations on WORD, DWORD and QWORD sizes don't work

### ISSUE THREAD

**Issue by datasynergyuk** [jovibor/HexCtrl#35]
To reproduce:

1. Open sample application
2. Fill with random writable data and then zero all data
3. Open data operations
4. Select Add: 0x1234
5. Select data size: Word

Nothing happens

It appears Word-Qword data operations don't work as expected

**Comment by jovibor** [comment]
I'm not sure i understand.
It works perfectly fine as expected, as always was. Can't reproduce neither in `Release` nor in `Debug`.

**Comment by datasynergyuk** [comment]
I have re-tested it and figured out what is going on. The problem is that the Operations dialog only modifies the current selection. As a user, if you click on a single byte, right click and then use the operations dialog, the expectation is naturally that all data will be modified. This is because it isn't clear to the user (well me) that in fact I have made a selection of 1-byte without even realising it. I would suggest an improvement to avoid this would be to add two new radio buttons to the dialog and auto-select the second option if a selection was present. This would make the workflow much clearer to the user:

a. All data
b. Selection (disabled if no selection present)

In other words, without realising it, I got confused by the 1-byte selection problem we have previously discussed.

This isn't a bug. Just a feature that isn't at all obvious unless you know how.

**Comment by jovibor** [comment]
>The problem is that the Operations dialog only modifies the current selection.

The problem? 0_o

I'm speechless here....

Of course it works with the selection. What else it would have been supposed to work with?! This is the primary purpose of the selection isn't it?!
You want modify all data? No problem, just select it all `Ctrl+A`.

>As a user, if you click on a single byte, right click and then use the operations dialog, the expectation is naturally that all data will be modified.

I don't know what the users you are talking about, but in no universe my expectations would be that ALL the data would be modified, it's just a nonsense. Could you please show any Hex editor that works that way? 
As for me it's just beyond common sense, and would be really bad UI design.

One byte is a one byte, you can perform any of Operations on exactly one byte (or, xor, add, etc...) can't you?

**Comment by jovibor** [comment]
Will close now as not an issue.

**Comment by datasynergyuk** [comment]
> Could you please show any Hex editor that works that way?

I have checked three different popular hex editors. I found all of them allow hex operations on both the current selection and the whole document. In one case (WinHex) this is implicit. In the case of the other two, the feature is a specific radio button in the UI. To me, this is common sense and HexCtrl is the outlier because it only performs hex operations on the selection and doesn't make this clear in the UI. 

I also looked at how each application handles single byte selections. To my surprise, each application does allow this (like HexCtrl) but unlike HexCtrl, all of them distinguish between moving the caret and making a one byte selection. In each case, the user must explicitly make the selection whilst in HexCtrl this happens implicitly. 

Combing the two points, this leads to the odd situation that the HexCtrl user typically always has (at least one byte) a selection and the operations dialog only works on this selection (but doesn't make this clear). In summary, I would suggest the following changes:

a. Moving the caret should not result in a single byte selection - Making a selection should be a positive act by the user. Moving the caret alone could clear the current selection or leave it in place.

b. Operations dialog should understand if a selection is present and offer the user the choice of whole file or selection.

Please understand, I make these suggestions in the spirit of wanting to improve HexCtrl and make it more powerful and user friendly. I don't want to get into a flame war about this.

Please find my detailed results below:

**WinHex**

WinHex allows single byte selection and supports operations on the selection (if present) or the whole file (if no selection). This is implicit and there is no UI to change this:

![image](https://user-images.githubusercontent.com/48149668/106353830-4d5e7180-62e5-11eb-8ebb-91d8d16b9317.png)

**Hex Workshop**

Hex Workshop allows single byte selection and supports operations on the selection (if present) or the whole file (if no selection). The UI offers the user the choice:

![image](https://user-images.githubusercontent.com/48149668/106353843-6535f580-62e5-11eb-93e6-351e1592343c.png)

010 Editor allows single byte selection and supports operations on the selection (if present) or the whole file (if no selection). The UI offers the user the choice:

**010 Editor**

![image](https://user-images.githubusercontent.com/48149668/106353849-71ba4e00-62e5-11eb-82e1-85584b33ec32.png)

**Comment by jovibor** [comment]
> **WinHex**
> 
> WinHex allows single byte selection and supports operations on the selection (if present) or the whole file (if no selection). This is implicit and there is no UI to change this:

Yes, you're right. Indeed, winhex works that way. But in my opinion it's absolutely incorrect.
You *implicitly* allows the user to amend the whole document. What if it's a memory, or really big file without undo? For my taste it's a-no-excuse.

As for the others, what's the point of the "Entire file" switch if you can simply select a whole document explicitly? 
What is the reason for overcomplicating things?...

**Comment by datasynergyuk** [comment]
A compromise would be to add two radio buttons to the operations dialog for:

1. Whole file
2. Selection (checked by default if selection present)

This would offer both features and make it explicit to the user what was happening. 

Please don't forget my comment about implicit single byte selection. I know you don't agree with me on it, but to my mind, HexCtrl's current behaviour is odd and leads to unintended consequences because of the implicit one-byte selection. For me, selections should always be a explicit act that is obvious to the user.
